apiserver/raft.go

```diff
@@ -0,0 +1,59 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiserver
+
+import (
+	"github.com/juju/errors"
+
+	apiservererrors "github.com/juju/juju/apiserver/errors"
+	"github.com/juju/juju/worker/raft"
+	"github.com/juju/juju/worker/raft/queue"
+)
+
+type Logger interface {
+	Debugf(string, ...interface{})
+	Tracef(string, ...interface{})
+}
+
+// Queue is a blocking queue to guard access and to serialize raft applications,
+// allowing for client side backoff.
+type Queue interface {
+	// Enqueue will add an operation to the queue. As this is a blocking queue, any
+	// additional enqueue operations will block and wait for subsequent operations
+	// to be completed.
+	// The design of this is to ensure that people calling this will have to
+	// correctly handle backing off from enqueueing.
+	Enqueue(queue.Operation) error
+}
+
+// raftMediator encapsulates raft related capabilities to the facades.
+type raftMediator struct {
+	queue  Queue
+	logger Logger
+}
+
+// ApplyLease attempts to apply the command on to the raft FSM. It only takes a
+// command and enqueues that against the raft instance. If the raft instance is
+// already processing a application, then back pressure is applied to the
+// caller and a ErrEnqueueDeadlineExceeded will be sent. It's up to the caller
+// to retry or drop depending on how the retry algorithm is implemented.
+func (m *raftMediator) ApplyLease(cmd []byte) error {
+	m.logger.Tracef("Applying Lease with command %s", string(cmd))
```

Comment:
IIRC a cast from `[]byte` to `string` creates a copy of the byte slice. If that's the case, and since this is a hot code path, we will be allocating *a lot* of objects on the heap even though the logger might not even be logging at TRACE level. 

My suggestion would be to use a wrapper that implements fmt.Stringer which in theory should be allocated on the stack:
```
type sliceStringer []byte
func (s sliceStringer) String(){ return string(s) }

//...
m.logger.Tracef("Applying Lease with command %s", sliceStringer(cmd))
```

Reply:
A better approach is to use `IsTraceEnabled` as that's standardised in Juju.